File: Easy/3442_Maximum_Difference_Between_Even_and_Odd_Frequency_I/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int maxDifference(string s) {
        unordered_map<char, int> freq;

        for (int i = 0; i < s.size(); i++) {
            freq[s[i]]++;
        }

        int max_odd = 0;
        int min_evn = INT_MAX;

        for (auto const& [val, count] : freq) {
            if (count % 2 != 0) {
                max_odd = max(max_odd, count);
            } else {
                min_evn = min(min_evn, count);
            }
        }

        return max_odd - min_evn;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    cout << solution.maxDifference("aaaaabbc") << endl;     // 3

    // test cases 2
    cout << solution.maxDifference("abcabcab") << endl;     // 1

    return 0;
}

File: Easy/3442_Maximum_Difference_Between_Even_and_Odd_Frequency_I/Solution.py
from typing import List, Optional


class Solution:
    def maxDifference(self, s: str) -> int:
        freq = {}

        for i in s:
            freq[i] = freq.get(i, 0) + 1

        odd_freq = []
        even_freq = []

        for count in freq.values():
            if count % 2 == 0:
                even_freq.append(count)
            else:
                odd_freq.append(count)

        return max(odd_freq) - min(even_freq)


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    s1 = "aaaaabbc"
    print(solution.maxDifference(s1))  # Output: 3
    
    # test cases 2
    s2 = "abcabcab"
    print(solution.maxDifference(s2))  # Output: 1
    
    